L1Trigger/TrackFindingTMTT/src/KFParamsComb.cc

```diff
@@ -0,0 +1,465 @@
+
+    double vphi(0);
+    double vz(0);
+    double vcorr(0);
+
+    // consider error due to integerisation only for z (r in encap) coord when enabled
+    double err_digi2(0);
+    if (getSettings()->enableDigitize())
+      err_digi2 = 0.15625 * 0.15625 / 12.0;
+
+    double a = stubCluster->sigmaX() * stubCluster->sigmaX();
+    double b = stubCluster->sigmaZ() * stubCluster->sigmaZ() + err_digi2;
+    double r2 = stubCluster->r() * stubCluster->r();
+    double invr2 = 1. / r2;
+
+    // Scattering term scaling as 1/Pt.
+    double sigmaScat = getSettings()->kalmanMultiScattTerm() / (state->candidate().pt());
+    double sigmaScat2 = sigmaScat * sigmaScat;
+
+    if (stubCluster->barrel()) {
+      vphi = (a * invr2) + sigmaScat2;
+
+      if (stubCluster->tiltedBarrel()) {
+        // Convert uncertainty in (r,phi) to (z,phi).
+        float scaleTilted = 1.;
+        if (getSettings()->kalmanHOtilted()) {
+          if (getSettings()->useApproxB()) {  // Simple firmware approximation
+            scaleTilted = getApproxB(stubCluster->z(), stubCluster->r());
+          } else {  // Exact C++ implementation.
+            float tilt = stubCluster->moduleTilt();
+            scaleTilted = sin(tilt) + cos(tilt) * tanl;
+          }
+        }
+        float scaleTilted2 = scaleTilted * scaleTilted;
+        // This neglects the non-radial strip effect, assumed negligeable for PS.
+        vz = b * scaleTilted2;
+      } else {
+        vz = b;
+      }
+
+      if (getSettings()->kalmanHOdodgy()) {
+        // Use original (Dec. 2016) dodgy implementation was this.
+        vz = b;
+      }
+
+    } else {
+      vphi = a * invr2 + sigmaScat2;
+      vz = (b * tanl2);
+
+      if (not stubCluster->psModule()) {  // Neglect these terms in PS
+        double beta = 0.;
+        // Add correlation term related to conversion of stub residuals from (r,phi) to (z,phi).
+        if (getSettings()->kalmanHOprojZcorr() == 2)
+          beta += -inv2R;
+        // Add alpha correction for non-radial 2S endcap strips..
+        if (getSettings()->kalmanHOalpha() == 2)
+          beta += -stubCluster->alpha();  // alpha is 0 except in endcap 2S disks
+
+        double beta2 = beta * beta;
+        vphi += b * beta2;
+        vcorr = b * (beta * tanl);
+
+        // IRT - for checking efficiency of removing phi-z correlation from projection.
+        // "ultimate_off1"
+        //vphi  = a * invr2 + b * pow(-stubCluster->alpha(), 2) + b * inv2R2 + sigmaScat2;
+        //vcorr = b * ((-stubCluster->alpha()) * tanl);
+
+        // IRT - This higher order correction doesn't significantly improve the track fit performance, so commented out.
+        //if (getSettings()->kalmanHOhelixExp()) {
+        //  float dsByDr = 1. + (1./2.)*r2*inv2R2; // Allows for z = z0 + s*tanL, where s is not exactly r due to circle.
+        //  vcorr *= dsByDr;
+        //  vz *= dsByDr * dsByDr;
+        //}
+
+        if (getSettings()->kalmanHOdodgy()) {
```

Comment:
Unfortunately, it corresponds to the current firmware, so is needed when comparing with that. The defaults used for production disable this option, since we are confident this can be fixed in firmware in future.

Reply:
Maybe it could have a better name than "dodgy"?

Reply:
Changed to kalmanHOfw (where fw is short for firmware).